File: src/main/java/info/movito/themoviedbapi/TmdbPeople.java
package info.movito.themoviedbapi;

import info.movito.themoviedbapi.model.movies.changes.ChangeResults;
import info.movito.themoviedbapi.model.people.ExternalIds;
import info.movito.themoviedbapi.model.people.PersonDb;
import info.movito.themoviedbapi.model.people.PersonImages;
import info.movito.themoviedbapi.model.people.Translations;
import info.movito.themoviedbapi.model.people.credits.CombinedPersonCredits;
import info.movito.themoviedbapi.model.people.credits.MovieCredits;
import info.movito.themoviedbapi.model.people.credits.TvCredits;
import info.movito.themoviedbapi.tools.ApiUrl;
import info.movito.themoviedbapi.tools.TmdbException;
import info.movito.themoviedbapi.tools.appendtoresponse.PersonAppendToResponse;

/**
 * The movie database api for people. See the
 * <a href="https://developer.themoviedb.org/reference/person-details">documentation</a> for more info.
 */
public class TmdbPeople extends AbstractTmdbApi {
    protected static final String TMDB_METHOD_PERSON = "person";

    /**
     * Create a new TmdbPeople instance to call the people related TMDb API methods.
     */
    TmdbPeople(TmdbApi tmdbApi) {
        super(tmdbApi);
    }

    /**
     * <p>Query the top level details of a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-details">documentation</a> for more info.</p>
     *
     * @param personId         The TMDb id of the person.
     * @param language         nullable - The language to query the results in. Default: en-US.
     * @param appendToResponse nullable - additional namespaces to append to the result (20 max).
     * @return The person details.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public PersonDb getDetails(int personId, String language, PersonAppendToResponse... appendToResponse) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId);
        apiUrl.addLanguage(language);
        apiUrl.addAppendToResponses(appendToResponse);
        return mapJsonResult(apiUrl, PersonDb.class);
    }

    /**
     * <p>Get the recent changes for a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-changes">documentation</a> for more info.</p>
     *
     * @param personId  The TMDb id of the person.
     * @param startDate nullable - The start date, in format: YYYY-MM-DD.
     * @param endDate   nullable - The end date, in format: YYYY-MM-DD.
     * @param page      nullable - The page of results to return. Default: 1.
     * @return The person changes.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public ChangeResults getChanges(int personId, String startDate, String endDate, Integer page) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "changes");
        apiUrl.addQueryParam("start_date", startDate);
        apiUrl.addQueryParam("end_date", endDate);
        apiUrl.addPage(page);
        return mapJsonResult(apiUrl, ChangeResults.class);
    }

    /**
     * <p>Get the combined movie and TV credits that belong to a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-combined-credits">documentation</a> for more info.</p>
     *
     * @param personId The TMDb id of the person.
     * @param language nullable - The language to query the results in. Default: en-US.
     * @return The combined person credits.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public CombinedPersonCredits getCombinedCredits(int personId, String language) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "combined_credits");
        apiUrl.addLanguage(language);
        return mapJsonResult(apiUrl, CombinedPersonCredits.class);
    }

    /**
     * <p>Get the external ID's that belong to a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-external-ids">documentation</a> for more info.</p>
     *
     * @param personId The TMDb id of the person.
     * @return The external IDs.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public ExternalIds getExternalIds(int personId) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "external_ids");
        return mapJsonResult(apiUrl, ExternalIds.class);
    }

    /**
     * <p>Get the profile images that belong to a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-images">documentation</a> for more info.</p>
     *
     * @param personId The TMDb id of the person.
     * @return The profile images.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public PersonImages getImages(int personId) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "images");
        return mapJsonResult(apiUrl, PersonImages.class);
    }

    /**
     * <p>Get the newest created person. This is a live response and will continuously change.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-latest-id">documentation</a> for more info.</p>
     *
     * @return The latest person.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public PersonDb getLatest() throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, "latest");
        return mapJsonResult(apiUrl, PersonDb.class);
    }

    /**
     * <p>Get the movie credits for a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-movie-credits">documentation</a> for more info.</p>
     *
     * @param personId The TMDb id of the person.
     * @param language nullable - The language to query the results in. Default: en-US.
     * @return The movie credits.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public MovieCredits getMovieCredits(int personId, String language) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "movie_credits");
        apiUrl.addLanguage(language);
        return mapJsonResult(apiUrl, MovieCredits.class);
    }

    /**
     * <p>Get the TV credits that belong to a person..</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/person-tv-credits">documentation</a> for more info.</p>
     *
     * @param personId The TMDb id of the person.
     * @param language nullable - The language to query the results in. Default: en-US.
     * @return The TV credits.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public TvCredits getTvCredits(int personId, String language) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "tv_credits");
        apiUrl.addLanguage(language);
        return mapJsonResult(apiUrl, TvCredits.class);
    }

    /**
     * <p>Get the translations that belong to a person.</p>
     * <p>See the <a href="https://developer.themoviedb.org/reference/translations">documentation</a> for more info.</p>
     *
     * @param personId The TMDb id of the person.
     * @return The translations.
     * @throws TmdbException If there was an error making the request or mapping the response.
     */
    public Translations getTranslations(int personId) throws TmdbException {
        ApiUrl apiUrl = new ApiUrl(TMDB_METHOD_PERSON, personId, "translations");
        return mapJsonResult(apiUrl, Translations.class);
    }
}

File: src/main/java/info/movito/themoviedbapi/TmdbPeople.kt
package info.movito.themoviedbapi

import info.movito.themoviedbapi.model.Artwork
import info.movito.themoviedbapi.model.ArtworkType
import info.movito.themoviedbapi.model.MovieImages
import info.movito.themoviedbapi.model.core.ResultsPage
import info.movito.themoviedbapi.model.people.Person
import info.movito.themoviedbapi.model.people.PersonCredits
import info.movito.themoviedbapi.model.people.PersonPeople
import info.movito.themoviedbapi.tools.ApiUrl
import info.movito.themoviedbapi.tools.MovieDbException


class TmdbPeople internal constructor(tmdbApi: TmdbApi) : AbstractTmdbApi(tmdbApi) {

    val DEFAULT_LANGUAGE = "en-US"
    /**
     * Get the latest person id.
     */
    val personLatest: PersonPeople
        get() {
           return personLatest(DEFAULT_LANGUAGE)
        }

    fun personLatest(language: String): PersonPeople {
        val apiUrl = ApiUrl(TMDB_METHOD_PERSON, "latest")

        apiUrl.addLanguage(language)

        return mapJsonResult(apiUrl, PersonPeople::class.java)
    }

    /**
     * This method is used to retrieve all of the basic person information.
     *
     *
     * It will return the single highest rated profile image.
     *
     * @param personId
     */



    fun getPersonInfo(personId: Int, vararg appendToResponse: String): PersonPeople{
        return getPersonInfo(DEFAULT_LANGUAGE, personId, *appendToResponse)
    }

    fun getPersonInfo(language: String, personId: Int, vararg appendToResponse: String): PersonPeople {
        val apiUrl = ApiUrl(TMDB_METHOD_PERSON, personId)

        apiUrl.appendToResponse(*appendToResponse)
        apiUrl.addLanguage(language)

        return mapJsonResult(apiUrl, PersonPeople::class.java)
    }

    /**
     * This method is used to retrieve all of the cast & crew information for the person.
     *
     *
     * It will return the single highest rated poster for each movie record.
     *
     * @param personId
     */

    fun getCombinedPersonCredits(personId: Int): PersonCredits {
        return getCombinedPersonCredits(personId, DEFAULT_LANGUAGE)
    }
    fun getCombinedPersonCredits(personId: Int, language: String): PersonCredits {
        val apiUrl = ApiUrl(TMDB_METHOD_PERSON, personId, "combined_credits")

        apiUrl.addLanguage(language)

        return mapJsonResult(apiUrl, PersonCredits::class.java)
    }


    /**
     * This method is used to retrieve all of the profile images for a person.
     *
     * @param personId
     */
    fun getPersonImages(personId: Int): List<Artwork> {
        val apiUrl = ApiUrl(TMDB_METHOD_PERSON, personId, "images")

        return mapJsonResult<MovieImages>(apiUrl, MovieImages::class.java).getAll(ArtworkType.PROFILE)
    }


    /**
     * Get the changes for a specific person id.
     *
     *
     * Changes are grouped by key, and ordered by date in descending order.
     *
     *
     * By default, only the last 24 hours of changes are returned.
     *
     *
     * The maximum number of days that can be returned in a single request is 14.
     *
     *
     * The language is present on fields that are translatable.
     *
     * @param personId
     * @param startDate
     * @param endDate
     */
    fun getPersonChanges(personId: Int, startDate: String, endDate: String) {
        throw MovieDbException("Not implemented yet")
    }


    /**
     * Get the list of popular people on The Movie Database.
     *
     *
     * This list refreshes every day.
     *
     * @param page
     * @return
     */

    fun getPersonPopular(page: Int?): PersonResultsPage {
        return getPersonPopular(page, DEFAULT_LANGUAGE)
    }

    fun getPersonPopular(page: Int?, language: String) : PersonResultsPage {
        val apiUrl = ApiUrl(TMDB_METHOD_PERSON, "popular")

        apiUrl.addPage(page)
        apiUrl.addLanguage(language)

        return mapJsonResult(apiUrl, PersonResultsPage::class.java)
    }



    class PersonResultsPage : ResultsPage<Person>()

    companion object {

        val TMDB_METHOD_PERSON = "person"
    }
}
